### PR TITLE
Fix and work around some issues on HIP/Crusher

### PR DESCRIPTION
**Pull request by sethrj**
With these workarounds/fixes, the release build works on Crusher. The debug build does not. I haven't dug into the ROOT failure but have disabled it for now since we can pull data directly from Geant4.

## UNRESOLVED: debug kernels are too complicated for HIP

```
[141/440] Building HIP object src/celeritas/CMakeFiles/celeritas.dir/em/generated/LivermorePEInteract.cu.o
FAILED: src/celeritas/CMakeFiles/celeritas.dir/em/generated/LivermorePEInteract.cu.o
/opt/rocm-5.1.0/llvm/bin/clang++ -DJSON_DIAGNOSTICS=0 -DJSON_USE_IMPLICIT_CONVERSIONS=1 -D__HIP_ROCclr__=1 -Dceleritas_EXPORTS -I/ccs/home/s3j/.local/src/celeritas/src -I/ccs/home/s3j/.local/src/celeritas/build/include -isystem /sw/crusher/spack-envs/base/opt/cray-sles15-zen3/gcc-11.2.0/nlohmann-json-3.10.2-fwiijvdxcf6lf7dfnojyswds4podhsem/include -munsafe-fp-atomics -Wno-#warnings -g -ggdb -O --offload-arch=gfx90a -fPIC -mllvm -amdgpu-early-inline-all=true -mllvm -amdgpu-function-calls=false -std=c++14 -MD -MT src/celeritas/CMakeFiles/celeritas.dir/em/generated/LivermorePEInteract.cu.o -MF src/celeritas/CMakeFiles/celeritas.dir/em/generated/LivermorePEInteract.cu.o.d -o src/celeritas/CMakeFiles/celeritas.dir/em/generated/LivermorePEInteract.cu.o -x hip -c /ccs/home/s3j/.local/src/celeritas/src/celeritas/em/generated/LivermorePEInteract.cu
error: branch size exceeds simm16
1 error generated when compiling for gfx90a.
```

## Debug assertions are broken in newer hip-clang

It seems like the hip-clang compiler validates the host/global function calls even outside the "device" pass (in which `__HIP_DEVICE_COMPILE__` is defined and the behavior of `CELER_ASSERT` changes).

```
FAILED: test/CMakeFiles/corecel_data_Collection.dir/corecel/data/Collection.test.cu.o
/opt/rocm-5.1.0/llvm/bin/clang++ -DGTEST_LINKED_AS_SHARED_LIBRARY=1 -D__HIP_ROCclr__=1 -I/ccs/home/s3j/.local/src/celeritas/test -I/ccs/home/s3j/.local/src/celeritas/build/test -I/ccs/home/s3j/.local/src/celeritas/src -I/ccs/home/s3j/.local/src/celeritas/build/include -isystem /sw/crusher/spack-envs/base/opt/linux-sles15-x86_64/gcc-7.5.0/googletest-1.10.0-75d47iea3s33viymgxd4goyvq5af3lmm/include -munsafe-fp-atomics -Wno-#warnings -g -ggdb -O --offload-arch=gfx90a -mllvm -amdgpu-early-inline-all=true -mllvm -amdgpu-function-calls=false -std=c++14 -MD -MT test/CMakeFiles/corecel_data_Collection.dir/corecel/data/Collection.test.cu.o -MF test/CMakeFiles/corecel_data_Collection.dir/corecel/data/Collection.test.cu.o.d -o test/CMakeFiles/corecel_data_Collection.dir/corecel/data/Collection.test.cu.o -x hip -c /ccs/home/s3j/.local/src/celeritas/test/corecel/data/Collection.test.cu
/ccs/home/s3j/.local/src/celeritas/test/corecel/data/Collection.test.cu:39:5: error: no matching constructor for initialization of '::celeritas::DebugError'
    CELER_ASSERT(matid);
    ^~~~~~~~~~~~~~~~~~~
/ccs/home/s3j/.local/src/celeritas/src/corecel/Assert.hh:190:32: note: expanded from macro 'CELER_ASSERT'
                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/ccs/home/s3j/.local/src/celeritas/src/corecel/Assert.hh:154:19: note: expanded from macro 'CELER_DEBUG_ASSERT_'
            throw ::celeritas::DebugError({::celeritas::DebugErrorType::WHICH, \
                  ^                       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/ccs/home/s3j/.local/src/celeritas/src/corecel/Assert.hh:435:7: note: candidate constructor (the implicit copy constructor) not viable: call to __host__ function from __global__ function
class DebugError : public std::logic_error
      ^
/ccs/home/s3j/.local/src/celeritas/src/corecel/Assert.hh:435:7: note: candidate constructor (the implicit move constructor) not viable: call to __host__ function from __global__ function
/ccs/home/s3j/.local/src/celeritas/src/corecel/Assert.hh:439:14: note: candidate constructor not viable: call to __host__ function from __global__ function
    explicit DebugError(DebugErrorDetails);
             ^
```

## missing `inline` keyword for FORCEINLINE macro
```
ld.lld: error: duplicate symbol: celeritas::ParticleTrackView::mass() const
>>> defined at ParticleProcessDiagnostic.cu
>>>            app/CMakeFiles/demo-loop.dir/demo-loop/diagnostic/ParticleProcessDiagnostic.cu.o:(celeritas::ParticleTrackView::mass() const)
>>> defined at StepDiagnostic.cu
>>>            app/CMakeFiles/demo-loop.dir/demo-loop/diagnostic/StepDiagnostic.cu.o:(.text+0x100)
```

## root@6.24.06 no longer finds the dictionary

@pcanal lots of other ROOT tests failed as well... looks like this is because the library with the root dictionary is now static, not shared, so it doesn't get loaded

```
ROOT: error: <TTree::Branch>:  The pointer specified for ImportData is not of a class or type known to ROOT
/ccs/home/s3j/.local/src/celeritas/app/celer-export-geant.cc:140: critical: Runtime error: celeritas: runtime error: failed to initialize ROOT ImportData
```

### CODE REVIEW

**Comment by sethrj** [comment]
Ping @tmdelellis now that you're back from boston :)

**Comment by pcanal** [comment]
The ROOT error seems to indicate that the dictionary are not being loaded (eg shared library not loaded, "unusused symbol" being discarded, etc..)

**Comment by sethrj** [comment]
Yeah @pcanal I think it's because the build is static—this changed in #526 .

File: src/corecel/Assert.hh [hunk 1]
```diff
@@ -180,7 +180,7 @@
     } while (0)
 //! \endcond
 
-#if CELERITAS_DEBUG && CELER_DEVICE_COMPILE
+#if CELERITAS_DEBUG && (CELER_DEVICE_COMPILE || defined(__HIP__))
```
Comment (tmdelellis) [inline src/corecel/Assert.hh:183]: Doesn't this use HIP automatically through `CELER_DEVICE_COMPILE`?
```cpp
#if defined(__CUDA_ARCH__) || defined(__HIP_DEVICE_COMPILE__)
#define CELER_DEVICE_COMPILE 1
#else
...
```

File: src/corecel/Assert.hh [hunk 2]
```diff
@@ -488,7 +488,7 @@ class RuntimeError : public std::runtime_error
 // INLINE FUNCTION DEFINITIONS
 //---------------------------------------------------------------------------//
 
-#if CELER_DEVICE_COMPILE
+#if CELER_DEVICE_COMPILE || defined(__HIP__)
```
Comment (tmdelellis) [inline src/corecel/Assert.hh:491]: Same as above

**Review by tmdelellis — COMMENTED** [review]
Just a few questions.

File: src/corecel/Assert.hh (hunk 1 above)
Comment (tmdelellis) [reply src/corecel/Assert.hh:183]: Ok, I guess this is a munged ramification of the fact that `HIP-Clang` does a single pass, so its not enough that this is a device compile, but that it is potentially using HIP to build host-side code for this translation unit as well, so you are further restricting this to only apply during the device-side compilation?

Comment (tmdelellis) [reply src/corecel/Assert.hh:183]: And actually, I just read your overview in more detail where you address this specifically.

**Review by tmdelellis — APPROVED** [review]

File: src/corecel/Assert.hh (hunk 1 above)
Comment (sethrj) [reply src/corecel/Assert.hh:183]: > clang uses merged parsing. This is similar to split compilation, except all of the host and device code is present and must be semantically-correct in both compilation steps.

So in clang when building host code you can't have device code present that throws, which is different from nvcc which entirely removes device code from the host pass.

**Pull request merged**